### DISCUSSION TITLE
can't figure out why formula / pkg will not install `elmer.rb`

### DISCUSSION THREAD

**Discussion by ipatch** [Homebrew/discussions#1877] [Tap maintenance and Homebrew development] [answered]
I've been working with the [homebrew freecad tap][1]. One of the tap maintainers checked in [`elmer.rb`][3] recently in the tap because [elmer][2] can be used in conjunction with freecad. I've reached out the tap maintainer, and he seems to be a bit preoccupied at the moment, thus me asking the questions here.

Basically, I'm attempting to get all formula files to pass `brew style` and `brew audit` to get the tap a couple of steps closer to generating auto bottles using ci/cd with github actions. I've managed to get all formula file to pass `brew style` but `elmer`.

I've taken a look at the elmer project page and github repo, and have managed to build and install elmer from the git source without any issues. i've managed to build the standalone cli version along with the gui version of elmer using the standard git clone, cmake, make, make install workflow.

However, when i attempt to perform the same steps (recipe) for elmer installing via the formula file using `brew` i keep running into errors.

It appears the issue seems to be qt related, which i'm a bit confused because i'm using the same cmake args from my shell via cli, as i'm defining within the formula file.

to install the cli only version of elmer

```
mkdir /some/path/on/fs/elmer
cd /some/path/on/fs/elmer
git clone https://github.com/ElmerCSC/elmerfem elmer-git
mkdir {build.cli,install.cli}
cd build.cli
cmake -DWITH_ELMERGUI:BOOL=FALSE -DWITH_MPI:BOOL=FALSE -DCMAKE_INSTALL_PREFIX=../install.cli ../elmer-git
make; say "build task done"
make install; say "install task done"
```

all of those steps were gathered from watching [this][4] youtube video as instructed by a elmer community member from the elmer youtube channel. the channel also hosts a video for building and install elmer gui which i'm also able to do. i just don't understand why when i try and run,

```
brew install freecad/freecad/elmer -v --head
```

i'm running into configuration issues with cmake,

i'm working with a modified version of the original formula below & [here][5]

any assistance with how i could streamline the edit formula file then rerunning `brew install ...` would greatly be appreciated. 🙏

[1]: <https://github.com/freecad/homebrew-freecad>
[2]: <https://github.com/ElmerCSC/elmerfem>
[3]: <https://github.com/FreeCAD/homebrew-freecad/blob/master/Formula/elmer.rb>
[4]: <https://www.youtube.com/watch?v=rQhSYKHw3HM>
[5]: <https://github.com/ipatch/homebrew-freecad/blob/07704d24f26e83985297f17cce540db461bca1ac/Formula/elmer.rb>

**Comment by ipatch** [comment]
running

```
brew install -i freecad/freecad/elmer -v --head
```

allowed me to poke around a bit with the environment and tweak some settings a little more interactively. 👍

**Comment by gromgit** [comment, marked as answer]
You didn't post the _exact_ error you saw (there's a reason why **Show, Don't Tell** is a rule of thumb when asking for technical help), but I'm guessing your CMake build config issues are due to this issue:
```
    args = ""
    args << "-DWITH_ELMERGUI:BOOL=FALSE "
    args << "-DWITH_QT5:BOOL=OFF "
    args << "-DWITH_MPI:BOOL=FALSE "
    args << "-DWITH_LUA:BOOL=FALSE "
    args << "-DWITH_Mumps:BOOL=FALSE "
```
You're building a `args` _string_ here, so this:
```
      system "cmake", *args, ".."
```
splats `args` into...a single string. That tells CMake to set the `WITH_ELMERGUI` boolean variable to the bizarre value:
```
FALSE -DWITH_QT5:BOOL=OFF -DWITH_MPI:BOOL=FALSE -DWITH_LUA:BOOL=FALSE -DWITH_Mumps:BOOL=FALSE 
```
including a final space at the end. I hardly ever use CMake, but I'm guessing that sets `WITH_ELMERGUI` to boolean _true_ (since it's not the exact string `FALSE`), and _doesn't_ set all the other  `WITH_` variables you're expecting.

Take a hint from every other core formula that builds command arguments incrementally, and use an _array_ instead:
```
    args = []
    args << "-DWITH_ELMERGUI:BOOL=FALSE"    # note: no trailing space
    ...
```
or better yet, since you don't seem to have any conditional logic:
```
    args = %w[
      -DWITH_ELMERGUI:BOOL=FALSE
      -DWITH_QT5:BOOL=OFF
      -DWITH_MPI:BOOL=FALSE
      -DWITH_LUA:BOOL=FALSE
      -DWITH_Mumps:BOOL=FALSE
    ]
```
then `*args` correctly splats the array into individual arguments for `cmake` to consume.
